### DISCUSSION TITLE
Installation on WSL2

### DISCUSSION THREAD

**Discussion by ic2010** [eliranwong/UniqueBible#1477] [General]
Hello.  Thank you for your wonderful project.  I was able to install UBA on my old system (Debian).  I have had to get a new laptop and am trying to install on WSL2 / Ubuntu or Debian (have tried both).  But, despite following the instructions, I either get a message saying the packages cannot be found or that there are issues with dependencies.  Is there a simple set of instructions that I could cut and paste to get it working on WSL2.  Sorry, a newbie.

**Comment by eliranwong** [comment]
I use Ubuntu on a PC and Debian on a Chrome OS. The app works very well.  May I ask any errors message displayed?

**Reply by ic2010** [reply]
Thank you very much for the reply!
I'm using Windows 11, WSL2 and the default Linux distribution from using wsl --install (Ubuntu 24.04.1 LTS at the moment).

I then followed the instructions on the page: https://github.com/eliranwong/UniqueBible/wiki/installation.
Installation looked like it went okay.

But, when I ran uniquebible, it tried installing the required files, but then crashed.  I do not have the expertise to interpret the problem, so I have attached output in the textfile.

Is that enough to tell you what I am doing wrong?  Thanks again in advance.

[ErrorFile.txt](https://github.com/user-attachments/files/17959701/ErrorFile.txt)

**Comment by eliranwong** [comment]
I guess you missed some library files.  For running Ubuntu on WSL2, please try to follow the instructions given on this page:

https://github.com/eliranwong/UniqueBible/wiki/Install-UBA-on-Windows-via-WSL2-Ubuntu

Please do install library files given at:

https://github.com/eliranwong/UniqueBible/wiki/Install-UBA-on-Windows-via-WSL2-Ubuntu#install-basic-library-and-tools

**Comment by ic2010** [comment]
Thank you again.  This was the output from: sudo apt -y install software-properties-common dirmngr apt-transport-https lsb-release ca-certificates apt-utils build-essential cmake tree wget curl git zip unzip xz-utils nano micro w3m lynx lib32stdc++6 sqlite3 libsqlite3-dev libasound2 libnss3 libncurses5 libncurses5-dev libgl1-mesa-dev mesa-utils libglu1-mesa lsb-release binutils ffmpeg gawk opencc plocate gnome-keyring libssl-dev libffi-dev libstdc++5 libpci3 libpci-dev python3 python3-setuptools python3-pip python3-dev python3-venv zlib1g-dev libgdbm-dev libnss3-dev libreadline-dev libbz2-dev gcc xorg-dev exo-utils dex gthumb nautilus xdg-utils

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Note, selecting 'libncurses-dev' instead of 'libncurses5-dev'
Package libasound2 is a virtual package provided by:
  liboss4-salsa-asound2 4.2-build2020-1ubuntu3
  libasound2t64 1.2.11-1build2 (= 1.2.11-1build2)
You should explicitly select one to install.

E: Unable to locate package lib32stdc++6
E: Package 'libasound2' has no installation candidate
E: Unable to locate package libncurses5
E: Unable to locate package libstdc++5

I came upon this with Ubuntu 24, 22, 20 and 18.  Is there a workaround?

**Reply by eliranwong** [reply]
Oh, sorry for the trouble, please try this one instead:

https://github.com/eliranwong/MultiAMDGPU_AIDev_Ubuntu/blob/main/ubuntu_desktop/basic.md#basic-tools--libraries

**Reply by eliranwong** [reply]
> Oh, sorry for the trouble, please try this one instead:
> 
> https://github.com/eliranwong/MultiAMDGPU_AIDev_Ubuntu/blob/main/ubuntu_desktop/basic.md#basic-tools--libraries

I tested it on my tablet running ubuntu 24

**Reply by ic2010** [reply]
Thank you for much for persisting with me... 

1. The command completed without issue: "sudo apt -y install software-properties-common dirmngr apt-transport-https lsb-release ca-certificates apt-utils build-essential make cmake tree wget curl git zip unzip xz-utils nano micro w3m lynx sqlite3 libsqlite3-dev sqlitebrowser libnss3 libnss3-dev libgl1-mesa-dev mesa-utils libglu1-mesa lsb-release binutils ffmpeg gawk opencc plocate gnome-keyring libssl-dev libffi-dev libpci3 libpci-dev python3 python3-setuptools python3-pip python3-dev python3-venv zlib1g-dev libgdbm-dev libreadline-dev libbz2-dev gcc xorg-dev exo-utils dex xdg-utils libavcodec-extra libportaudio2 moreutils llvm tk-dev liblzma-dev python3-openssl libxml2-dev libxmlsec1-dev protobuf-compiler libc6-dev libstdc++-12-dev libxcb-cursor-dev libxcb-cursor0 libncurses-dev libncurses6 ubuntu-restricted-addons ubuntu-restricted-extras gstreamer1.0-plugins-base gstreamer1.0-plugins-good gstreamer1.0-plugins-bad gstreamer1.0-plugins-ugly xsel portaudio19-dev vlc python3-wheel python3-wheel-whl twine"

2. Followed the "Upgrade cmake" section and entered "curl https://pyenv.run/ | bash" and edited the ".bashrc" file.

3. But, "pyenv install --list | grep 3.10.1" and "pyenv install 3.10.14" both gave me

Command 'pyenv' not found, did you mean:
  command 'p7env' from deb libnss3-tools (2:3.96.1-1)
Try: sudo apt install <deb name>" 
 pyenv install --list | grep 3.10.1

4. When I tried running uniquebible, I would get...
Traceback (most recent call last):
  File "/home/ian/UniqueBible/venv/lib/python3.12/site-packages/uniquebible/main.py", line 5, in <module>
    from uniquebible.util.FileUtil import FileUtil
ModuleNotFoundError: No module named 'uniquebible'

Apologies... I thinking I'm progressing, but still not able to run the program :)

**Comment by eliranwong** [comment]
Sorry, I was not clear about the instructions.  The step 1 you mentioned is enough, pyenv is optional.

Please try the following steps:

1. check if there is a `uniquebible` folder in directory, `~/apps`.  If so, delete it, to make it clean for a fresh installation to rule out other factors:

> rm -rf ~/apps/uniquebible

It would be fine if it does not exists.

2. Copy the all the following lines and paste into terminal and run (press Enter):

```
cd ~
mkdir -p apps
cd apps
python3 -m venv uniquebible
source uniquebible/bin/activate
pip install --upgrade uniquebible
echo "alias uniquebible='"$(which uniquebible)"'" >> ~/.bashrc
source ~/.bashrc
uniquebible
```

Please kindly wait for all instructions to run through ...

Please kindly let me know if it works.

**Reply by eliranwong** [reply]
> Sorry, I was not clear about the instructions. The step 1 you mentioned is enough, pyenv is optional.
> 
> Please try the following steps:
> 
> 1. check if there is a `uniquebible` folder in directory, `~/apps`.  If so, delete it, to make it clean for a fresh installation to rule out other factors:
> 
> > rm -rf ~/apps/uniquebible
> 
> It would be fine if it does not exists.
> 
> 2. Copy the all the following lines and paste into terminal and run (press Enter):
> 
> ```
> cd ~
> mkdir -p apps
> cd apps
> python3 -m venv uniquebible
> source uniquebible/bin/activate
> pip install --upgrade uniquebible
> echo "alias uniquebible='"$(which uniquebible)"'" >> ~/.bashrc
> source ~/.bashrc
> uniquebible
> ```
> 
> Please kindly wait for all instructions to run through ...
> 
> Please kindly let me know if it works.

If it works, you only need to run the following word in the terminal to call the program:

> uniquebible

**Reply by eliranwong** [reply]
> Sorry, I was not clear about the instructions. The step 1 you mentioned is enough, pyenv is optional.
> 
> Please try the following steps:
> 
>     1. check if there is a `uniquebible` folder in directory, `~/apps`.  If so, delete it, to make it clean for a fresh installation to rule out other factors:
> 
> 
> > rm -rf ~/apps/uniquebible
> 
> It would be fine if it does not exists.
> 
>     2. Copy the all the following lines and paste into terminal and run (press Enter):
> 
> 
> ```
> cd ~
> mkdir -p apps
> cd apps
> python3 -m venv uniquebible
> source uniquebible/bin/activate
> pip install --upgrade uniquebible
> echo "alias uniquebible='"$(which uniquebible)"'" >> ~/.bashrc
> source ~/.bashrc
> uniquebible
> ```
> 
> Please kindly wait for all instructions to run through ...
> 
> Please kindly let me know if it works.

If you previously followed the instructions above, please run the app, by entering one word `uniquebible`

If it does not work, can you please run cat `~/.bashrc` and shows me the last few lines of the output?

**Comment by ic2010** [comment]
Thank you for your ongoing patience!
Got stuck on another dependency issue :)

[libwebp.txt](https://github.com/user-attachments/files/17963582/libwebp.txt)

**Comment by eliranwong** [comment]
Please try to install Qt6, described in the first part of https://github.com/eliranwong/MultiAMDGPU_AIDev_Ubuntu/blob/main/ubuntu_desktop/applications/development/qt.md

![Screenshot_20241130-003333](https://github.com/user-attachments/assets/b271e6ed-2ef0-41a9-b281-85c27c50621b)

Please try to run `uniquebible` afterwards and let me know.  Thanks.

**Reply by ic2010** [reply]
Sorry... I'm like that student who can't past the exam!!!

Tried the above and got a similar error message as before...

Sorry.  I know you've got plenty of commitments on yours hands, so please don't feel obliged to keep troubleshooting my problem :)

ic2010@LAPTOP:~/apps$ uniquebible
Setting up user directory 'audio' ...
Setting up user directory 'htmlResources' ...
Setting up user directory 'import' ...
Setting up user directory 'macros' ...
Setting up user directory 'marvelData' ...
Setting up user directory 'music' ...
Setting up user directory 'notes' ...
Setting up user directory 'temp' ...
Setting up user directory 'terminal_history' ...
Setting up user directory 'terminal_mode' ...
Setting up user directory 'thirdParty' ...
Setting up user directory 'video' ...
Setting up user directory 'webstorage' ...
Setting up user directory 'workspace' ...
ic2010@LAPTOP:~/apps$ pip tool updated!
Installing 'PySide6' ...
Module 'PySide6' is installed!
Installed!
Installing 'html-text' ...
Module 'html-text' is installed!
Installed!
Installing 'html5lib' ...
Module 'html5lib' is installed!
Installed!
Installing 'mammoth' ...
Module 'mammoth' is installed!
Installed!
Installing 'htmldocx' ...
Module 'htmldocx' is installed!
Installed!
Installing 'diff_match_patch' ...
Module 'diff_match_patch' is installed!
Installed!
Installing 'langdetect' ...
Module 'langdetect' is installed!
Installed!
Installing 'pygithub' ...
Module 'pygithub' is installed!
Installed!
Installing 'qt-material' ...
Module 'qt-material' is installed!
Installed!
Installing 'telnetlib3' ...
Module 'telnetlib3' is installed!
Installed!
Installing 'ibm-watson' ...
Module 'ibm-watson' is installed!
Installed!
Installing 'translate' ...
Module 'translate' is installed!
Installed!
Installing 'qrcode' ...
Module 'qrcode' is installed!
Installed!
Installing 'pillow' ...
Module 'pillow' is installed!
Installed!
Installing 'yt-dlp' ...
Module 'yt-dlp' is installed!
Installed!
Installing 'gTTS' ...
Module 'gTTS' is installed!
Installed!
Installing 'markdownify' ...
Module 'markdownify' is installed!
Installed!
Installing 'nltk' ...
Module 'nltk' is installed!
Installed!
Installing 'word-forms' ...
Module 'word-forms' is installed!
Installed!
Installing 'lemmagen3' ...
Module 'lemmagen3' is installed!
Installed!
Installing 'chinese-english-lookup' ...
Module 'chinese-english-lookup' is installed!
Installed!
Installing 'tabulate' ...
Module 'tabulate' is installed!
Installed!
Installing 'pyluach' ...
Module 'pyluach' is installed!
Installed!
Installing 'gmplot' ...
Module 'gmplot' is installed!
Installed!
Installing 'art' ...
Module 'art' is installed!
Installed!
Installing 'colorama' ...
Module 'colorama' is installed!
Installed!
Installing 'numpy' ...
Module 'numpy' is installed!
Installed!
Installing 'matplotlib' ...
Module 'matplotlib' is installed!
Installed!
Installing 'asyncssh' ...
Module 'asyncssh' is installed!
Installed!
Installing 'bcrypt' ...
Module 'bcrypt' is installed!
Installed!
Installing 'validators' ...
Module 'validators' is installed!
Installed!
Installing 'pydub' ...
Module 'pydub' is installed!
Installed!
Installing 'SpeechRecognition' ...
Module 'SpeechRecognition' is installed!
Installed!
Installing 'pocketsphinx' ...
Module 'pocketsphinx' is installed!
Installed!
Installing 'googlesearch-python' ...
Module 'googlesearch-python' is installed!
Installed!
Installing 'guidance' ...
Module 'guidance' is installed!
Installed!
Installing '--upgrade gdown' ...
Module '--upgrade gdown' is installed!
Downloading initial content 'images.sqlite' ...
Downloaded!
Unpacking ...
'images.sqlite' is installed!
Downloading initial content 'cCBSC.commentary' ...
Downloaded!
Unpacking ...
'cCBSC.commentary' is installed!
Error in sys.excepthook:
Traceback (most recent call last):
  File "/home/ic2010/apps/uniquebible/lib/python3.12/site-packages/uniquebible/startup/share.py", line 47, in global_excepthook
    print(traceback.format_exc())
          ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'traceback' object has no attribute 'format_exc'

Original exception was:
Traceback (most recent call last):
  File "/home/ian/apps/uniquebible/lib/python3.12/site-packages/uniquebible/main.py", line 125, in <module>
    from uniquebible.startup.guiQt import *
  File "/home/ian/apps/uniquebible/lib/python3.12/site-packages/uniquebible/startup/guiQt.py", line 16, in <module>
    from uniquebible.gui.MainWindow import MainWindow
  File "/home/ian/apps/uniquebible/lib/python3.12/site-packages/uniquebible/gui/MainWindow.py", line 15, in <module>
    from PySide6.QtWebEngineCore import QWebEnginePage
ImportError: libwebp.so.6: cannot open shared object file: No such file or directory

**Comment by eliranwong** [comment]
From the error message, it appears that the library libwebp.so.6 is missing.

I search my computer for this file by running locate `libwebp.so.6`

```
/snap/codium/404/usr/lib/x86_64-linux-gnu/libwebp.so.6
/snap/codium/404/usr/lib/x86_64-linux-gnu/libwebp.so.6.0.2
/snap/codium/428/usr/lib/x86_64-linux-gnu/libwebp.so.6
/snap/codium/428/usr/lib/x86_64-linux-gnu/libwebp.so.6.0.2
```

It appears that the file is installed with codium, so I would suggest you to try again by installing codium:

> sudo snap install codium

Please kindly let me know if it works after installing codium, thanks

**Reply by ic2010** [reply]
Thank you for persevering with me...
This is the error message from running uniquebible

Traceback (most recent call last):
  File "/home/ian/apps/uniquebible/lib/python3.12/site-packages/uniquebible/main.py", line 5, in <module>
    from uniquebible.util.FileUtil import FileUtil
ModuleNotFoundError: No module named 'uniquebible'

Thank you for all your help so far.

**Reply by eliranwong** [reply]
How do you run the app?

**Reply by ic2010** [reply]
I load up the Ubuntu environment via the Windows Start Menu.
Once within the Ubuntu environment, I type uniquebible via the command line interface.  Is that what you mean?

**Comment by eliranwong** [comment]
Okay, I finally got a Windows to test it, and found the issue, please use the revised instructions here.

To install on Windows via WSL:

https://github.com/eliranwong/UniqueBible/wiki/Install-UBA-on-Windows-via-WSL2-Ubuntu#install-uba

To install on Windows directly:

https://github.com/eliranwong/UniqueBible/wiki/Install-on-Windows

**Reply by ic2010** [reply]
Thanks again.  I think I must have a problem with how I'm setting things up.

I loaded a Powershell as an administrator.
Unregistered all previous Linux distributions to start from scratch.
Then wsl --install and cut and pasted the commands from the "Install basic library and tools" section.  No issues.
The commands from "Install UBA" ran without problems.
But, uniquebible came up with the messages below.
I did "sudo snap install --classic codium", a command similar to an earlier post, but the error persisted.

Was there anything different I should have been doing?



Error in sys.excepthook:
Traceback (most recent call last):
  File "/home/ian/ub/lib/python3.12/site-packages/uniquebible/startup/share.py", line 47, in global_excepthook
    print(traceback.format_exc())
          ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'traceback' object has no attribute 'format_exc'

Original exception was:
Traceback (most recent call last):
  File "/home/ian/ub/bin/ubgui", line 8, in <module>
    sys.exit(gui())
             ^^^^^
  File "/home/ian/ub/lib/python3.12/site-packages/uniquebible/uba.py", line 157, in gui
    main()
  File "/home/ian/ub/lib/python3.12/site-packages/uniquebible/uba.py", line 151, in main
    exec("from uniquebible.main import *", globals())
  File "<string>", line 1, in <module>
  File "/home/ian/ub/lib/python3.12/site-packages/uniquebible/main.py", line 125, in <module>
    from uniquebible.startup.guiQt import *
  File "/home/ian/ub/lib/python3.12/site-packages/uniquebible/startup/guiQt.py", line 16, in <module>
    from uniquebible.gui.MainWindow import MainWindow
  File "/home/ian/ub/lib/python3.12/site-packages/uniquebible/gui/MainWindow.py", line 15, in <module>
    from PySide6.QtWebEngineCore import QWebEnginePage
ImportError: libwebp.so.6: cannot open shared object file: No such file or directory

**Reply by eliranwong** [reply]
may I ask if you use the revised instructions at https://github.com/eliranwong/UniqueBible/wiki/Install-UBA-on-Windows-via-WSL2-Ubuntu#install-uba to install UBA?

**Reply by ic2010** [reply]
Yes 😃 that's why I'm wondering if it's how I've set it up because even starting with a new wsl --install gave me the same error 😔

**Reply by ic2010** [reply]
I've tried the instructions on someone else's laptop.  Windows 11 / WSL2 / Ubuntu 24
It works flawlessly.
Tried it again on mine, same problem as above.
My laptop is an ARM device (Snapdragon X) while the friend's one is an older Intel i7.  I wonder if that is the reason for it not working on mine?

Sigh...

**Reply by eliranwong** [reply]
> I've tried the instructions on someone else's laptop. Windows 11 / WSL2 / Ubuntu 24 It works flawlessly. Tried it again on mine, same problem as above. My laptop is an ARM device (Snapdragon X) while the friend's one is an older Intel i7. I wonder if that is the reason for it not working on mine?
> 
> Sigh...

Sorry, I don't have a Snapdragon device to check.

**Reply by ic2010** [reply]
It's all good.  Thank you heaps for all your time and effort.  Much appreciated.
